triggers/daml/Daml/Trigger.daml

```diff
@@ -260,3 +268,109 @@ data TriggerState s = TriggerState
   , commandsInFlight : Map CommandId [Command]
   , nextCommandId : Int
   }
+
+-- | Used to construct an 'ACS' for 'testRule'.
+newtype ACSBuilder = ACSBuilder [Update (AnyContractId, AnyTemplate)]
+
+instance Semigroup ACSBuilder where
+  ACSBuilder l <> ACSBuilder r = ACSBuilder (l <> r)
+
+instance Monoid ACSBuilder where
+  mempty = ACSBuilder mempty
+
+buildACS : Party -> ACSBuilder -> Scenario ACS
+buildACS party (ACSBuilder fetches) = do
+  activeContracts <- submit party $ sequence fetches
+  pure ACS
+    { activeContracts = activeContracts
+    , pendingContracts = Map.empty
+    }
+
+-- | Include the given contract in the 'ACS'.
+toACS : Template t => ContractId t -> ACSBuilder
+toACS cid = ACSBuilder
+  [fetch cid >>= \tpl -> pure (toAnyContractId cid, toAnyTemplate tpl)]
+
+-- | Execute a trigger's rule once in a scenario.
+testRule
+  : Trigger s  -- ^ Test this trigger's 'Trigger.rule'.
+  -> Party  -- ^ Execute the rule as this 'Party'.
+  -> ACSBuilder  -- ^ List these contracts in the 'ACS'.
+  -> Map CommandId [Command]  -- ^ The commands in flight.
+  -> s  -- ^ The trigger state.
+  -> Scenario [Commands]  -- ^ The 'Commands' emitted by the rule. The 'CommandId's will start from @"0"@.
+testRule trigger party acsBuilder commandsInFlight s = do
+  time <- getTime
+  acs <- buildACS party acsBuilder
+  let state = TriggerState
+        { acs = acs
+        , party = party
+        , userState = s
+        , commandsInFlight = commandsInFlight
+        , nextCommandId = 0
+        }
+  let (_, commands) = runRule trigger.rule time state
+  pure commands
+
+-- | Drop 'CommandId's and extract all 'Command's.
+flattenCommands : [Commands] -> [Command]
+flattenCommands = concatMap commands
+
+expectCommand
+  : [Command]
+  -> (Command -> Optional a)
+  -> (a -> Either Text ())
+  -> Either [Text] ()
+expectCommand commands fromCommand assertion = foldl step (Left []) commands
+  where
+    step : Either [Text] () -> Command -> Either [Text] ()
+    step (Right ()) _ = Right ()
+    step (Left msgs) command =
+      case assertion <$> fromCommand command of
+        None -> Left msgs
+        Some (Left msg) -> Left (msg :: msgs)
+        Some (Right ()) -> Right ()
+
+-- | Check that at least one command fulfills the given assertions.
+assertCreateCmd
```

Comment:
With these assertion helpers it is easy to check that a given command was emitted but it is tricky to assert that a command was not emitted or that a given command was emitted multiple times (although I don’t really see a good usecase for the latter). Not sure how important this is so let’s definitely leave it for a separate PR but worth keeping in mind.

Reply:
Yes, agreed. I'd say this is a first draft at the assertion API. We'll have to see what API is demanded as use-cases come up.